### ISSUE TITLE
floats: add SameScalars to compare two float64 values with NaN == NaN.

### ISSUE THREAD

**Issue by romanwerpachowski** [gonum/gonum#1405]
**What**

`floats/floats.go` has a function

```
// Same returns true if the input slices have the same length and all elements
// have the same value with NaN treated as the same.
func Same(s, t []float64) bool {
...}
````

I would like to add there a function `SameScalars` which does the same for scalar inputs:

```
// SameScalars returns true if the inputs have the same value with NaN treated as the same.
func SameScalars(a, b float64) bool {
	return a == b || (math.IsNaN(a) && math.IsNaN(b))
}
```

**Rationale**

Such a function will be useful for unit testing functions which are expected to return NaNs for some inputs.

**Updated proposal**

Rename `Same` to `AllSame` (breaks code) and call the new function `Same`.

**Comment by kortschak** [comment]
A counter naming proposal would be to rename `Same` to `AllSame` and call the new function `Same`.

**Comment by romanwerpachowski** [comment]
That makes sense, but what about backwards compatibility? Do we know how much code outside Gonum would this renaming break?

**Comment by kortschak** [comment]
It's something that we need to consider, but we are prepared to break people for a better API while we haven't stabilised the API (we are pre-v1).

**Comment by kortschak** [comment]
This is a comment for #1407, but here because this is an API discussion.

For consistency, I would suggest (and this is why this needs more discussion) changes for the slice equality functions. The current
```
func Equal(s1, s2 []float64) bool
func EqualApprox(s1, s2 []float64, tol float64) bool
func EqualFunc(s1, s2 []float64, f func(float64, float64) bool) bool
```
become
```
func AllEqual(s1, s2 []float64) bool
func AllEqualApprox(s1, s2 []float64, tol float64) bool
func AllEqualFunc(s1, s2 []float64, f func(float64, float64) bool) bool
```
which will obviously break many people.

This is much more invasive, but I think also more indicative of what is happening.

Unfortunately, the package name is overloaded. The equivalent packages in the standard library are bytes (and by extension strings where a `string` is effectively a `const []byte`). The bytes package is like the floats package parts that deal with `[]float64`. We also have floats doing things with the single element. This means that the `bytes.Equal` model doesn't work since we are not now necessarily saying are the collections of ~bytes~ floats equal in all cases (the various approximate equality tests that we have don't make it clear from the name that they are talking about single float values).

/cc @btracey @vladimir-ch @sbinet

**Comment by romanwerpachowski** [comment]
The current API is already inconsistent, because `EqualApprox` takes slices, but `EqualWithinAbs` takes single values.

**Comment by kortschak** [comment]
Yes, that's part of what I suggesting be fixed.

**Comment by romanwerpachowski** [comment]
Ah, right. Sorry.

There isn't that much time till August, so others' comments are more then welcome :)

**Comment by kortschak** [comment]
The real time is a week or so after the release of Go1.15, so probably after August.

**Comment by romanwerpachowski** [comment]
Great. So, I personally am for renaming everything now to make things consistent. Better than half-measures.

**Comment by btracey** [comment]
I think we should keep `Equal` as is, and either move or rename the ones involving scalars. As you mention, the package is `floats` and has the core type of `[]float64`. `EqualWithinAbs` and friends were mostly added to support `EqualApprox`. I don't really see why we would have `AllEqual`, when all of the other functions implicitly apply to all of the elements. I would recommend something like having a `floats/scalar` package, where we put `SameScalar` (now `scalar.Same`), as well as `EqualWithinAbs`,  `NanWith`, etc.

That keeps floats as dealing with `[]float64`, and has a new pacakge that conceptually deals with "tricky scalar operations"

**Comment by kortschak** [comment]
Yeah, fair enough. The other option is to call it `float` though the spelling closeness to `floats` pretty much blocks that. `.../floats/scalar` works for me. There is a cognate split that has to happen for `cmplxs` which mirrors `floats`. Any suggestions for that?

**Comment by romanwerpachowski** [comment]
Won't the functions operating on scalars be used most often? They should be in the top level package, and the vector functions in the subpackage, IMHO.

**Comment by romanwerpachowski** [comment]
*ping*

**Comment by kortschak** [comment]
I'd argue that the vectory things are more commonly used.

@btracey Do you have a suggestion for a reasonably name for the cmplxs cognate of floats/scalar?

**Comment by romanwerpachowski** [comment]
OK, then let's have `floats` and `floats/scalar`. Do we need to fix the name for `cmplxs` before I can proceed with `floats`?

**Comment by kortschak** [comment]
I'd prefer that it all happen in one PR for catching consistency errors.

**Comment by btracey** [comment]
One idea is just to have `cmplxs/scalar` . It's not the best with goimports, but would be fine code wise. Another is to do cscalar. I think that the complex versions will be a lot less frequently used, so if we're going to choose it's better to make the floats one "nice"

**Comment by kortschak** [comment]
Yeah, I was thinking ".../cmplxs/scalar" or ".../cmplxs/cscalar" too. The only issue is that complex values aren't scalars; though I think maybe I just need to leave that pedantry behind. If we go this route, I would prefer "cscalar" since it catches bugs better.

**Comment by btracey** [comment]
Agreed with cscalar

**Comment by kortschak** [comment]
OK. I'll get to this later today.

**Comment by kortschak** [comment]
Closed by #1418.
